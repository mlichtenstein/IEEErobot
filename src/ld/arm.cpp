#include "Settings.h"
#include "arm.h"


Servo R1,R2_1,R2_2,R3;
int newArmTheta[] = {0,0,0};
int oldArmTheta[] = {0,0,0};

// arm function prototypes

double convertR1(int inDegree);
double convertR2_1(int inDegree);
double convertR2_2(int inDegree);
double convertR3(int inDegree);

void setupArm() {
  //write arm to a safe angle befor bringing it online
    R1.writeMicroseconds(convertR1(90));
    R2_1.writeMicroseconds(convertR2_1(0));
    R2_2.writeMicroseconds(convertR2_1(0));
    R3.writeMicroseconds(convertR3(90));
    
    R1.attach(36);
    R2_1.attach(37);
    R2_2.attach(38);
    R3.attach(39);
    pinMode(8, OUTPUT);
    delay(1000);
}
void armControl() {
    newArmTheta[2] = 275 - newArmTheta[2];
    for(int i = 1; i <= 200; i++) {
        //Serial.write( 'i' );
        R1.writeMicroseconds((int)(convertR1(oldArmTheta[0]) + (convertR1(newArmTheta[0]) - convertR1(oldArmTheta[0]))*i/200));
        R2_1.writeMicroseconds((int)(convertR2_1(oldArmTheta[1]) + (convertR2_1(newArmTheta[1]) - convertR2_1(oldArmTheta[1]))*i/200));
        R2_2.writeMicroseconds((int)(convertR2_2(oldArmTheta[1]) + (convertR2_2(newArmTheta[1]) - convertR2_2(oldArmTheta[1]))*i/200));
        R3.writeMicroseconds((int)(convertR3(oldArmTheta[2]) + (convertR3(newArmTheta[2]) - convertR3(oldArmTheta[2]))*i/200));
        delay(10);
    }
    for(int i = 0; i < 3; i++) {
        oldArmTheta[i] = newArmTheta[i];
    }

}
void magnet(char inChar)
{
   if(inChar == 'i')
     digitalWrite(8, HIGH);
   else if(inChar == 'o')
     digitalWrite(8,LOW);
}

void magnet(char inChar)
{
  Serial.println("farts");
  if(inChar == 'i')
   digitalWrite(8, HIGH);
  else if(inChar =='o')
    digitalWrite(8, LOW);
}

// Servo angle conversion and callibration functions
// retrun [offset] + (inputDegrees/360 * [scale]
// scale = the difference between the offset and the value
// that gives a full360 of rotation
double convertR1(int inDegree) {
    return 1080 + ((double)inDegree/360 * 2350);
}
double convertR2_1(int inDegree) {
    return 1005 + ((double)inDegree/360 * 2180);
}
double convertR2_2(int inDegree) {
    return 980 + ((double)(inDegree)/360 * 2430);
}
double convertR3(int inDegree) {
    return 650 + ((double)inDegree/360 * 1800);
}
