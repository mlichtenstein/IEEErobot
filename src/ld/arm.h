#if defined( ROBOT_SERVICE_ARM_SERVO ) && !defined( GUARD_ARM_HEADER )
#define GUARD_ARM_HEADER
extern int newArmTheta[];
void setupArm();
void armControl();
void magnet(char aChar);
#ifndef TESTING
#if ARDUINO >= 100
#include "Arduino.h"
#else
#include "WProgram.h"
#endif
#endif
#endif
