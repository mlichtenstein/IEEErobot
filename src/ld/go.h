#if defined( ROBOT_SERVICE_GO ) && !defined( GUARD_GO_HEADER )
#define GUARD_GO_HEADER
void scoot(double distance, double angle);

void go(double x, double y, double theta);
void turn(double angle);


void right(double Angle);

void left(double Angle);

void forward(double feet);

void reverse(double feet);
void motorSpeed(double inSpeed);

void wheelAngle(double FL_SERVO, double FR_SERVO);

void faceWheels(double angle);

void ackSolve(float theta, double motorSpeed);

void ackTest();
void setupGo() ;
#endif
