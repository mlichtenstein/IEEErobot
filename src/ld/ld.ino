#include "Settings.h"

#include "go.h"
#include "scan.h"
#include "arm.h"

#include <Servo.h>

/// A counter for out going IDs.
int messageID = 0;

char inBoxBuffer[ 1000 ] = "";
int inBoxBuffer_newChar = 0;

void establishContact( );
bool readMessage( );

void setup() {
    #ifdef ROBOT_WAIT_MODE
    #define startButton 12
    pinMode(startButton, INPUT);
    digitalWrite(startButton, HIGH);
    #endif
    #ifdef ROBOT_SERVICE_SCAN
    setupScan();
    #endif
    Serial.begin( ROBOT_SERIAL_PORT_SPEED );
    establishContact( );
    Serial.flush();
    delay( 100 );

    //le arm setup
    #ifdef ROBOT_SERVICE_ARM_SERVO
    setupArm();
    #endif
    #ifdef ROBOT_SERVICE_GO
    setupGo();
    #endif
}

/* Go Example
void loop()
{
        //go(0,1,0);go(2,0,0);
        forward(3); reverse(3);turn(90);turn(-90);scoot(10,45);
        while(1){int i = 0;}
}
*/

void loop() {
    if ( Serial.available() > 0 ) {
        if ( readMessage( ) ) {
            char type;
            int id;
            int numOfVars = sscanf( inBoxBuffer, "%c%d", &type, &id );
            if ( numOfVars == 2 ) {
                switch ( type ) {
                    /*  The following are where you paste your arduino code.  Make sure an entry in Settings.h corresponds to your case.
                    Your response message should take the form ":[char],[id],[payload];"
                    */

#ifdef ROBOT_PING_TEST
                case ROBOT_PING_TEST: {
                    Serial.write( ":P," );
                    Serial.print( id );
                    Serial.write(",PONG");
                    Serial.write( ';' );
                    Serial.write( ":C," );
                    Serial.print( id );
                    Serial.write( ';' );
                }
                break;
#endif
                //this block tests sending doubles to ard.
                case 'd': {
                    Serial.write(":d,");
                    Serial.print(id);
                    Serial.write(",");
                    int myDouble;
                    char spoil;
                    int numOfVars = sscanf( inBoxBuffer, "%c%d", &spoil, &myDouble );
                    Serial.print(myDouble);
                    Serial.print(';');
                    Serial.write( ":C," );
                    Serial.print( id );
                    Serial.write( ';' );
                }
                break;
                //this block is used for calibrating savox servos
#ifdef ROBOT_SERVICE_CALIBRATE_SERVO
                case ROBOT_SERVICE_CALIBRATE_SERVO: {
                    Serial.write( ":M" );
                    Serial.print( id );
                    Serial.write( ';' );
                    for (int i=0; i<4; i++) {
                        EyeServo[i].write(id);
                    }
                    Serial.write( ":C," );
                    Serial.print( id );
                    Serial.write( ';' );
                }
#endif
                //this block is used for testing the calibration of savox servos
#ifdef ROBOT_SERVICE_TEST_SERVO
                case ROBOT_SERVICE_TEST_SERVO: {
                    Serial.write( ":L" );
                    Serial.print( id );
                    Serial.write( ';' );
                    for (int i=0; i<4; i++) {
                        EyeServoWrite(i, id);
                    }
                    Serial.write( ":C," );
                    Serial.print( id );
                    Serial.write( ';' );
                }
                break;
#endif
#ifdef ROBOT_SERVICE_GO
                case ROBOT_SERVICE_GO: {
                    // Forward declarations.
                    char subCategory;
                    int temp;
                    boolean messageGood = false;
                    numOfVars = sscanf( inBoxBuffer, "%*c%*d,%c,%d", &subCategory, &temp );
                    float firstParam = (float)temp / 120.0; //Convert pixels to feet
                    Serial.write("Feet=");
                    Serial.print(firstParam);
                    // Should have received 2 parameters.
                    if ( numOfVars == 2 ) {
                        // Turn.
                        if ( subCategory == ROBOT_COMMAND_TURN ) {
                            if ( firstParam >= -360 && firstParam <= 360 ) {
                                turn( firstParam*120 ); //Uh I guess we wanted degrees not feet
                                messageGood = true;
                            }
                            // Bad angle.
#ifndef NODEBUG
                            else {
                                Serial.write( ':' );
                                Serial.write( ROBOT_RESPONSE_ERROR );
                                Serial.print( ROBOT_SERIAL_ERROR_WRONG_ARGUMENTS );
                                Serial.write( "Error: Bad angle of\"" );
                                Serial.print( firstParam );
                                Serial.write( "\" was greater than 360 or less than -360." );
                                Serial.write( "Message was\"" );
                                Serial.write( inBoxBuffer );
                                Serial.write( "\";" );
                            }
#endif
                        }
                        // forward fuck
                        else if ( subCategory == ROBOT_COMMAND_FORWARD ) {
                            forward( firstParam );
                            messageGood = true;

                        }
                        // Scoot.
                        else if ( subCategory == ROBOT_COMMAND_SCOOT ) {
                            numOfVars = sscanf( inBoxBuffer, "%*c%*d,%*c,%*d,%d", &temp );
                            float angle = (float)temp;
                            if ( numOfVars == 1 ) {
                                if ( angle >= -360 && angle <= 360 ) {
                                    scoot( firstParam, angle );
                                    messageGood = true;
                                }
                                // Bad angle.
#ifndef NODEBUG
                                else {
                                    Serial.write( ':' );
                                    Serial.write( ROBOT_RESPONSE_ERROR );
                                    Serial.print( ROBOT_SERIAL_ERROR_WRONG_ARGUMENTS );
                                    Serial.write( "Error: Bad angle of\"" );
                                    Serial.print( angle );
                                    Serial.write( "\" was greater than 360 or less than -360." );
                                    Serial.write( "Message was\"" );
                                    Serial.write( inBoxBuffer );
                                    Serial.write( "\";" );
                                }
#endif
                            }
                            // Wrong number of arguments to scoot. Should have got 2 but got only 1.
#ifndef NODEBUG
                            else {
                                Serial.write( ':' );
                                Serial.write( ROBOT_RESPONSE_ERROR );
                                Serial.print( ROBOT_SERIAL_ERROR_WRONG_ARGUMENTS );
                                Serial.write( "Error: Wrong number of arguments to scoot. Expecting 2 but 1. Message was\"" );
                                Serial.write( inBoxBuffer );
                                Serial.write( "\";" );
                            }
#endif
                        }
                        // Bad subcategory command.
#ifndef NODEBUG
                        else {
                            Serial.write( ':' );
                            Serial.write( ROBOT_RESPONSE_ERROR );
                            Serial.print( ROBOT_SERIAL_ERROR_WRONG_ARGUMENTS );
                            Serial.write( "Error: Bad sub category of \"." );
                            Serial.print( subCategory );
                            Serial.write( "\";" );
                            Serial.write( "Message was\"" );
                            Serial.write( inBoxBuffer );
                            Serial.write( "\";" );
                        }
#endif
                    }
                    // Wrong number of arguments.
#ifndef NODEBUG
                    else {
                        Serial.write( ':' );
                        Serial.write( ROBOT_RESPONSE_ERROR );
                        Serial.print( ROBOT_SERIAL_ERROR_WRONG_ARGUMENTS );
                        Serial.write( "Error: Wrong number of arguments to the go service. Expecting 1 but there was none. Message was\"" );
                        Serial.write( inBoxBuffer );
                        Serial.write( "\";" );
                    }
#endif
                    // Command succeeded.
                    if ( messageGood ) {
                        Serial.write( ":C," );
                        Serial.print( id );
                        Serial.write( ';' );
                    }
                }
                break;
#endif
#ifdef ROBOT_WAIT_MODE
                //this block is "wait mode," a simple poll for a button
                case ROBOT_WAIT_MODE: {
                    for (int i=0; i<4; i++) {
                        EyeServo[i].write(45);
                    }
                    while(digitalRead(startButton)==HIGH);
                    Serial.write(":W,");
                    Serial.print(id);
                    Serial.write(";");
                    Serial.write( ":C," );
                    Serial.print( id );
                    Serial.write( ';' );
                }
                break;
#endif
#ifdef ROBOT_SERVICE_SCAN
                case ROBOT_SERVICE_SCAN: {

                    float deltaTheta = ROBOT_SCAN_ANGLE/(ROBOT_SCAN_DATA_POINTS-1);
                    unsigned int IRreading[4][ROBOT_SCAN_DATA_POINTS];
                    unsigned int USreading[4][ROBOT_SCAN_DATA_POINTS];

                    for (int pt = 0; pt <= ROBOT_SCAN_DATA_POINTS; pt += 1) {
                        unsigned long lastTime = millis();  //used to establish a minimum read time
                        for (int eye = 0; eye<4; eye++) {
                            EyeServoWrite(eye,pt);
                            USreading[eye][pt] = PingFire(eye);  //this can be slow if we do it 4 times...might need more delicate code
                            IRreading[eye][pt] = analogRead(IRpin[eye]);
                        }
                        while (millis() <= lastTime+DELAY) {} //wait for minimum read time to elapse, if nec
                    }
                    //now send the message
                    Serial.write( ":J," );
                    Serial.print( id );
                    Serial.write( ",");
                    for (int pt = 0; pt <= ROBOT_SCAN_DATA_POINTS; pt += 1) {
                        for (int eye = 0; eye < 4; eye+=1) {
                            Serial.print( char('|'));
                            Serial.print( (eye)); //EyeNum
                            Serial.write( ',' );
                            Serial.print( (pt)); //DataPointNum
                            Serial.write( ',' );
                            Serial.print( (IRreading[eye][pt] & 255)); //IRlsb
                            Serial.write( ',' );
                            Serial.print( (IRreading[eye][pt] >> 8));  //IRmsb
                            Serial.write( ',' );
                            Serial.print( (USreading[eye][pt] & 255)); //USlsb
                            Serial.write( ',' );
                            Serial.print( (USreading[eye][pt] >> 8));  //USlsb
                        }
                    }
                    Serial.write( ';' );
                    for (int rundown = 180; rundown >= 0; rundown-=5){
                        for (int i = 0; i<4; i++) {
                            EyeServoWrite(i,rundown);
                            //                      EyeServo[i].write(0);
                            delay(10);
                        }

                    }
                    Serial.write( ":C," );
                    Serial.print( id );
                    Serial.write( ';' );
                }
                break;
#endif
#ifdef ROBOT_SERVICE_ARM_SERVO
                case ROBOT_SERVICE_ARM_SERVO: {
                    numOfVars = sscanf( inBoxBuffer, "%*c%*d,%d,%d,%d",
                                        &( newArmTheta[0] ), &( newArmTheta[1] ), &( newArmTheta[2] ) );
                    if ( numOfVars == 3 ) {
                        Serial.print( newArmTheta[0] );
                        Serial.write( ',' );
                        Serial.print( newArmTheta[1] );
                        Serial.write( ',' );
                        Serial.print( newArmTheta[2] );
                        Serial.write( ',' );

                        Serial.write( ':' );
                        Serial.write( ROBOT_RESPONSE_COMFIRM );
                        Serial.print( id );
                        Serial.write( ';' );
                        armControl();
                    } else {
                        Serial.write( ':' );
                        Serial.write( ROBOT_RESPONSE_ERROR );
                        Serial.print( ROBOT_SERIAL_ERROR_WRONG_ARGUMENTS );
                        Serial.write( "Error: Wrong number of arguments. Message was\"" );
                        Serial.write( inBoxBuffer );
                        Serial.write( "\";" );
                    }
                    Serial.write( ":C," );
                    Serial.print( id );
                    Serial.write( ';' );
                }
                break;
#endif
#ifndef NODEBUG
                default: {
                    Serial.write( ':' );
                    Serial.write( ROBOT_RESPONSE_ERROR );
                    Serial.print( ROBOT_SERIAL_ERROR_NO_SERVICE );
                    Serial.write( "Error: No such service. Message was\"" );
                    Serial.write( inBoxBuffer );
                    Serial.write( "\";" );

                }
#endif
                }
            }
#ifndef NODEBUG
            else {
                Serial.write( ':' );
                Serial.write( ROBOT_RESPONSE_ERROR );
                Serial.print( ROBOT_SERIAL_ERROR_BAD_MESSAGE );
                Serial.write( "Error: Bad message. Message was\"" );
                Serial.write( inBoxBuffer );
                Serial.write( "\";" );
            }
#endif
        }
    }
#ifdef HAS_ARM_SERVO
    if ( armManager.inMove() ) {
        for (int i = 0; i < ARM_SERVOS; ++i ) {
            int theta = armManager.currentTheta[ i ];
            if ( armManager.move( i, millis(), &theta ) ) {
                R[ i ].write( theta );
            }
        }
    }
#endif
    delay( 150 );
}

/**
 * Reads the bytes from the serial buffer and returns true
 *  when a full message is received in full.
 *
 * @param serial can take the "Serial" object.
 * @param outMessage is used to accumulate bytes from the buffer.
 *
 * @return TRUE if a message is received or FALSE otherwise.
 */
bool readMessage( ) {
    char chr;
    static bool foundMessage = false;
    while ( Serial.available() > 0 ) {
        chr = Serial.read();
        if ( chr == ROBOT_SERIAL_MESSAGE_START ) {
            foundMessage = true;
            inBoxBuffer_newChar = 0;
        } else if ( chr == ROBOT_SERIAL_MESSAGE_STOP ) {
            if ( inBoxBuffer_newChar != 0 ) {
                inBoxBuffer[ inBoxBuffer_newChar ] = '\0';
                return true;
            }
            foundMessage = false;
        } else if ( foundMessage )
            inBoxBuffer[ inBoxBuffer_newChar++ ] = chr;
    }
    inBoxBuffer[ inBoxBuffer_newChar ] = '\0';
    return false;
}

/**
 * Waits for a greeting then returns a greeting. This
 *  function could never terminate if a byte is not
 *  received.
 *
 * Preconditions:
 * -Serial.begin() has been called successfully.
 *
 * @param serial can take the "Serial" object.
 */
void establishContact( ) {
    while ( Serial.available() <= 0 ) {
        delay( 300 );
    }
    // Write back one char.
    Serial.write( Serial.read() );
}
