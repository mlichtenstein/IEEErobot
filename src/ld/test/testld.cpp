// Enable testing
#define TESTING
// For print.
#include <stdio.h>
// Testing macros.
#include "tests.h"
// For strlen()
#include <string.h>


// BEGIN Mock Declarations
#include <malloc.h>
#define boolean bool
#define abs( X ) ( (X) < 0 ? -(X) : (X) )
#define PI 3.14
#define A1 0
#define A2 1
#define A3 1
void delay( int ) {
}
void analogWrite( int, int ) {
}
int digitalRead( int ) {
}
class HardwareSerial {
  public:
    const char* buffer;
    int nextChar;
    char outBuffer[1000];
    int out_nextChar;
    HardwareSerial( )
            : nextChar( 0 ), out_nextChar( 0 ) {
        outBuffer[0] = '\0';
    }
    char read() {
        char ret = buffer[ nextChar ];
        if ( ret ) {
            ++nextChar;
        }
        return ret;
    }
    int available() {
        return strlen( buffer + nextChar );
    }
    void write( char out ) {
        outBuffer[ out_nextChar++ ] = out;
        outBuffer[ out_nextChar ] = '\0';
    }
    void write( const char* out ) {
        int len = strlen( out );
        strcpy( outBuffer + out_nextChar, out );
        out_nextChar += len;
    }
    void flush() {
    }
    void begin( int ) {
    }
    void print( int out ) {
        int len = sprintf( outBuffer + out_nextChar, "%d", out );
        out_nextChar += len;
    }
};
#include <Servo.h>
void Servo::write( int ) {
}
uint8_t Servo::attach( int ) {
}
Servo::Servo( ) {
}
HardwareSerial Serial;
#define A0 0
#define A1 0
#define OUTPUT 0x1
#define INPUT 0x0
#define LOW  0x0
#define HIGH 0x1
typedef unsigned char uint8_t;
void digitalWrite(uint8_t pin, uint8_t val) {
}
void pinMode(uint8_t, uint8_t) {
}
void delayMicroseconds(unsigned int us) {
}
unsigned long pulseIn(uint8_t pin, uint8_t state, unsigned long timeout = 1000000L) {
}
unsigned long millis(void) {
    static int time = 0;
    return time++;
}
int analogRead(uint8_t pin) {
}
// END Mock Declarations

// Pull in test file.
#include "../ld.ino"



// Forward declaration.
int goodBehavior_TestSuite();

/**
 * Main calls test_suite() and returns results.
 *
 * @return the number of failures.
 */
int main() {
  printf( "line %d\n", __LINE__ );
	int ret=goodBehavior_TestSuite();
	return ret;
}

// BEGIN Results
/// Holds the number of failed assertions.
int bad_asserts = 0;
/// Holds the number of total assertions.
int total_assertions = 0;
/// Holds the number of failed tests.
int bad_tests = 0;
/// Holds the number of total tests.
int total_tests = 0;
// END Results


/**
 * Proves decodeMessage will convert a string message into
 *  a decodedMessage in most cases.
 *
 * @return TRUE if the test succeeds.
 */
bool goodBehavior_loop() {
  bool status = true;
  ++total_tests;
  char type;
  int id;
  TEST_EQUAL( 2,  sscanf( "L123", "%c%d", &type, &id ) );
  TEST_EQUAL( 'L',  type );
  TEST_EQUAL( 123,  id );
  printf( "line %d\n", __LINE__ );
  // Check an empty message.
  {
      Serial.buffer = ":J123;";
      Serial.nextChar = 0;
      Serial.out_nextChar = 0;
      Serial.outBuffer[0] = '\0';
      const char* expected = ":C,123;";
  printf( "line %d\n", __LINE__ );
      loop();
  printf( "line %d\n", __LINE__ );
      char * temp = Serial.outBuffer;
      char * lastColon = (char*)-1;
      while ( *temp != '\0' ) {
        if ( *temp == ':' )
            lastColon = temp;
        ++temp;
      }
      TEST_( lastColon != (char*)-1,
            printf( "FAILURE BECAUSE: expected \"%s\" and actual was \"%s\".\n", expected, Serial.outBuffer ) );
      TEST_( strcmp( expected, lastColon ) == 0,
            printf( "FAILURE BECAUSE: expected \"%s\" and actual was \"%s\".\n", expected, lastColon ) );
      if ( !status ) {
        printf( "outbuffer: %s\n", Serial.outBuffer );
        printf( "type: %c\n", type );
        printf( "id: %d\n", id );
        printf( "lastColon: %s\n", lastColon );
      }
  }
  printf( "line %d\n", __LINE__ );

  // Check a bad service.
  {
      Serial.buffer = ":*123;";
      Serial.nextChar = 0;
      Serial.out_nextChar = 0;
      Serial.outBuffer[0] = '\0';
      loop();
      char type;
      int id;
      TEST_EQUAL( 6, Serial.nextChar );
      TEST_EQUAL( 2, sscanf( Serial.outBuffer, ":%c%d%*s;", &type, &id ) );
      TEST_EQUAL( ROBOT_RESPONSE_ERROR, type );
      TEST_EQUAL( ROBOT_SERIAL_ERROR_NO_SERVICE, id );
      if ( !status ) {
        printf( "outbuffer: %s\n", Serial.outBuffer );
        printf( "type: %c\n", type );
        printf( "id: %d\n", id );
        printf( "Serial.out_nextChar: %d\n", Serial.out_nextChar );
        printf( "Serial.nextChar: %d\n", Serial.nextChar );
      }
  }
  printf( "line %d\n", __LINE__ );

  // Check a bad message.
  {
      Serial.buffer = ":123;";
      Serial.nextChar = 0;
      Serial.out_nextChar = 0;
      Serial.outBuffer[0] = '\0';
      loop();
      char type;
      int id;
      TEST_EQUAL( 2, sscanf( Serial.outBuffer, ":%c%d%*s;", &type, &id ) );
      TEST_EQUAL( ROBOT_RESPONSE_ERROR, type );
      TEST_EQUAL( ROBOT_SERIAL_ERROR_BAD_MESSAGE, id );
      if ( !status ) {
        printf( "outbuffer: %s\n", Serial.outBuffer );
        printf( "type: %c\n", type );
        printf( "id: %d\n", id );
      }

      Serial.buffer = ":ta123;";
      Serial.nextChar = 0;
      loop();
      TEST_EQUAL( 2, sscanf( Serial.outBuffer, ":%c%d%*s;", &type, &id ) );
      TEST_EQUAL( ROBOT_RESPONSE_ERROR, type );
      TEST_EQUAL( ROBOT_SERIAL_ERROR_BAD_MESSAGE, id );
      if ( !status ) {
        printf( "outbuffer: %s\n", Serial.outBuffer );
        printf( "type: %c\n", type );
        printf( "id: %d\n", id );
      }
  }
  printf( "line %d\n", __LINE__ );

  // Check a incomplete message.
  {
      Serial.buffer = ":123";
      Serial.nextChar = 0;
      Serial.out_nextChar = 0;
      Serial.outBuffer[0] = '\0';
      loop();
      char type;
      int id;
      TEST_EQUAL( 2, sscanf( Serial.outBuffer, ":%c%d%*s;", &type, &id ) );
      TEST_EQUAL( ROBOT_RESPONSE_ERROR, type );
      TEST_EQUAL( ROBOT_SERIAL_ERROR_BAD_MESSAGE, id );
      if ( !status ) {
        printf( "outbuffer: %s\n", Serial.outBuffer );
        printf( "type: %c\n", type );
        printf( "id: %d\n", id );
      }

      Serial.buffer = "ta123;";
      Serial.nextChar = 0;
      Serial.outBuffer[0] = '\0';
      loop();
      TEST_EQUAL( 0, strlen( Serial.outBuffer ) );
      if ( !status ) {
        printf( "outbuffer: %s\n", Serial.outBuffer );
        printf( "type: %c\n", type );
        printf( "id: %d\n", id );
      }
  }
  printf( "line %d\n", __LINE__ );

  if ( status )
    printf( "Test %s successful!\n", __FUNCTION__ );
  return status;
}

/**
 * Proves decodeMessage will convert a string message into
 *  a decodedMessage in most cases.
 *
 * @return TRUE if the test succeeds.
 */
bool goodBehavior_readMessage() {
  bool status = true;
  ++total_tests;
  printf( "line %d\n", __LINE__ );
  // Check an empty message.
  {
      Serial.buffer = ":;";
      Serial.nextChar = 0;
      Serial.out_nextChar = 0;
      Serial.outBuffer[0] = '\0';
      TEST_EQUAL( false, readMessage() );
  }
  printf( "line %d\n", __LINE__ );

  if ( status )
    printf( "Test %s successful!\n", __FUNCTION__ );
  return status;
}

bool goodBehavior_confirmations() {
  bool status = true;
  ++total_tests;
  printf( "line %d\n", __LINE__ );
  // Check IR Scan confirmation.
  {
      char buffer[100];
      const char services[] = {
            ROBOT_PING_TEST,
            //ROBOT_SERVICE_CAMERA,
            //ROBOT_SERVICE_IMO,
            ROBOT_PING_TEST,
            ROBOT_SERVICE_SCAN,
            ROBOT_SERVICE_CALIBRATE_SERVO,
            ROBOT_SERVICE_TEST_SERVO,
            //ROBOT_SERVICE_ARM_SERVO,
            ROBOT_SERVICE_GO,
            ROBOT_CALIB_IR,
            ROBOT_WAIT_MODE
      };
  printf( "line %d\n", __LINE__ );
      int serviceCount = sizeof( services );
      for ( int i = 0; i < serviceCount; ++i ) {
  printf( "line %d\n", __LINE__ );
          sprintf( buffer, ":%c%d;", services[ i ], 1014 );
            printf( "input88: \"%s\"\n", buffer );
          Serial.buffer = buffer; //":J1014;";
          Serial.nextChar = 0;
          Serial.out_nextChar = 0;
  printf( "line %d\n", __LINE__ );
          const char* expected = ":C,1014;";
          loop();
  printf( "line %d\n", __LINE__ );
          char * temp = Serial.outBuffer;
          char * lastColon = (char*)-1;
          while ( *temp != '\0' ) {
            if ( *temp == ':' )
                lastColon = temp;
            ++temp;
          }
          TEST_( lastColon != (char*)-1,
                printf( "FAILURE BECAUSE: expected \"%s\" and actual was \"%s\".\n", expected, Serial.outBuffer ) );
          TEST_( strcmp( expected, lastColon ) == 0,
                printf( "FAILURE BECAUSE: expected \"%s\" and actual was \"%s\".\n", expected, lastColon ) );
          if ( !status ) {
            printf( "input: \"%s\"\n", buffer );
            printf( "output: \"%s\"\n", Serial.outBuffer );
            printf( "expected: \"%s\"\n", expected );
          }
      }
  }

  if ( status )
    printf( "Test %s successful!\n", __FUNCTION__ );
  return status;
}

/**
 * Runs "good" behavor tests.
 *
 * @return the number of failures
 */
int goodBehavior_TestSuite(){
  bad_asserts = 0;
  total_assertions = 0;
  printf( "line %d\n", __LINE__ );
  printf( "BEGIN Good Behavior Tests\n");
  printf( "line %d\n", __LINE__ );
  if ( !goodBehavior_confirmations() )
    ++bad_tests;
  if ( !goodBehavior_loop() )
    ++bad_tests;
  printf( "line %d\n", __LINE__ );

    // Run each test and count the successes.
  printf( "line %d\n", __LINE__ );


  printf( "END Good Behavior Tests\n");

    // Prints results.
  printf( "%d of %d assertions successful!\n", (total_assertions - bad_asserts), total_assertions );
  printf( "%d of %d tests successful!\n", (total_tests - bad_tests), total_tests );

    // Print status.
  if( bad_tests == 0 )
    printf("Success!\n");
  else
    printf("Failure!\n");

  return bad_tests;
}


