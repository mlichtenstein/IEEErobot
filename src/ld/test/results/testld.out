BEGIN Good Behavior Tests
Test goodBehavior_readMessage successful!
FAILURE BECAUSE: expected ":C123;" and actual was ":E1Error: No such service. Message was"J123";".
Actual: strcmp( expected, Serial.outBuffer ) == 0 = FALSE
Expected: True
Assertion failed!
More Info: File "test/testld.cpp"
Line: 133 Function: "goodBehavior_loop"
Expected: 2 = 2
Actual: sscanf( Serial.outBuffer, "%c%d%*s", &type, &id ) = 1
Assertion of "2 == sscanf( Serial.outBuffer, "%c%d%*s", &type, &id )" failed!
More Info: File "test/testld.cpp"
Line: 143 Function: "goodBehavior_loop"
Expected: ROBOT_RESPONSE_ERROR = 69
Actual: type = 58
Assertion of "ROBOT_RESPONSE_ERROR == type" failed!
More Info: File "test/testld.cpp"
Line: 144 Function: "goodBehavior_loop"
Expected: ROBOT_SERIAL_ERROR_NO_SERVICE = 1
Actual: id = 25
Assertion of "ROBOT_SERIAL_ERROR_NO_SERVICE == id" failed!
More Info: File "test/testld.cpp"
Line: 145 Function: "goodBehavior_loop"
Expected: 2 = 2
Actual: sscanf( Serial.outBuffer, "%c%d%*s", &type, &id ) = 1
Assertion of "2 == sscanf( Serial.outBuffer, "%c%d%*s", &type, &id )" failed!
More Info: File "test/testld.cpp"
Line: 155 Function: "goodBehavior_loop"
Expected: ROBOT_RESPONSE_ERROR = 69
Actual: type = 58
Assertion of "ROBOT_RESPONSE_ERROR == type" failed!
More Info: File "test/testld.cpp"
Line: 156 Function: "goodBehavior_loop"
Expected: ROBOT_SERIAL_ERROR_BAD_MESSAGE = 1
Actual: id = 25
Assertion of "ROBOT_SERIAL_ERROR_BAD_MESSAGE == id" failed!
More Info: File "test/testld.cpp"
Line: 157 Function: "goodBehavior_loop"
Expected: 2 = 2
Actual: sscanf( Serial.outBuffer, "%c%d%*s", &type, &id ) = 1
Assertion of "2 == sscanf( Serial.outBuffer, "%c%d%*s", &type, &id )" failed!
More Info: File "test/testld.cpp"
Line: 162 Function: "goodBehavior_loop"
Expected: ROBOT_RESPONSE_ERROR = 69
Actual: type = 58
Assertion of "ROBOT_RESPONSE_ERROR == type" failed!
More Info: File "test/testld.cpp"
Line: 163 Function: "goodBehavior_loop"
Expected: ROBOT_SERIAL_ERROR_BAD_MESSAGE = 1
Actual: id = 25
Assertion of "ROBOT_SERIAL_ERROR_BAD_MESSAGE == id" failed!
More Info: File "test/testld.cpp"
Line: 164 Function: "goodBehavior_loop"
Expected: 0 = 0
Actual: strlen( Serial.outBuffer ) = 176
Assertion of "0 == strlen( Serial.outBuffer )" failed!
More Info: File "test/testld.cpp"
Line: 174 Function: "goodBehavior_loop"
Expected: 0 = 0
Actual: strlen( Serial.outBuffer ) = 225
Assertion of "0 == strlen( Serial.outBuffer )" failed!
More Info: File "test/testld.cpp"
Line: 179 Function: "goodBehavior_loop"
END Good Behavior Tests
4 of 16 assertions successful!
1 of 2 tests successful!
Failure!
